cohortextractor/dsl.py

```diff
@@ -0,0 +1,309 @@
+
+    def first_for_patient(self) -> PatientFrame:
+        """Return a PatientFrame with the first_for_patient event for each patient."""
+
+        return PatientFrame(self, "first_for_patient")
+
+    def last_for_patient(self) -> PatientFrame:
+        """Return a PatientFrame with the last_for_patient event for each patient."""
+
+        return PatientFrame(self, "last_for_patient")
+
+    def _first_for_patient(
+        self,
+        table: QMTable,
+        state: State,
+    ) -> tuple[Row, State]:
+        return table.first_by(*state["sort_columns"]), {}
+
+    def _last_for_patient(
+        self,
+        table: QMTable,
+        state: State,
+    ) -> tuple[Row, State]:
+        return table.last_by(*state["sort_columns"]), {}
+
+
+class PatientFrame(Node):
+    """Represents a collection of records, with one row per patient.
+
+    Either a PatientTable, or the result of filtering a PatientTable.
+    """
+
+    def filter(self, column: str, **kwargs: str) -> PatientFrame:  # noqa: A003
+        """Return a new PatientFrame with given filter."""
+
+        return PatientFrame(self, "filter", column, **kwargs)
+
+    def select_column(self, column: str) -> PatientSeries:
+        """Return a PatientSeries containing given column."""
+
+        return PatientSeries(self, "select_column", column)
+
+    def _filter(
+        self,
+        table: QMTable,
+        state: State,
+        column: str,
+        **kwargs: str,
+    ) -> tuple[QMTable, State]:
+        if kwargs.get("not_equals", object()) is None:
+            assert "not_null_filter" not in state
+            extra_state = {"not_null_filter": column}
+        else:
+            extra_state = {}
+        return table.filter(column, **kwargs), extra_state
+
+    def _select_column(
+        self,
+        row: Row,
+        state: State,
+        column: str,
+    ) -> tuple[ValueFromRow, State]:
+        return row.get(column), {}
+
+
+class PatientSeries(Node):
+    """Represents a column indexed by patient.
+
+    Can be used as a variable in a Cohort, or as an input when computing another
+    variable.
+    """
+
+
+class BaseTable:
+    """A base class for database tables."""
+
+    def __init__(self, name):
+        self.name = name
+        self._path = []
+
+
+class EventTable(BaseTable, EventFrame):
+    """A base class for database tables with multiple rows per patient."""
+
+
+class PatientTable(BaseTable, PatientFrame):
+    """A base class for database tables with one row per patient."""
+
+
+def compile(cohort: Cohort) -> dict[str, Value]:  # noqa: A001
+    """Return a dict mapping variable names to Query Model values."""
+
+    return {k: compile_variable(v._path) for k, v in vars(cohort).items()}
+
+
+def compile_variable(path: list[PathItem]) -> Value:
+    """Return the Query Model value for this variable."""
+
+    obj: QueryNode = Table(path[0][0].name)
+    state: State = {}
+
+    for node, method_name, args, kwargs in path:
```

Comment:
Would it be clearer if  PathItem was a dataclass/namedtuple so we can use dotted attributes?  Because it's a named thing in the type annotation, I was expecting it to be used like a class instance and had to remind myself it was defined as a tuple

Reply:
Another good idea.